Performance Improvements (4x)

jakob opened the pull request:
I've added two basic performance tests for the parser (I wanted to compare the performance to our own parser).

I found some low hanging fruit and fixed it. The parser is now roughly 4 times faster on the test documents.

jakob committed:
Add Performance Tests

Added two tests to measure the performance of the CSV parser.

Two sample files have also been added for testing different scenarios:
- semi-big-file.csv is a sample file that contains many short rows
- postico-issues.csv is a sample file that contains a dump of Github issues. It has very long fields

jakob committed:
Improve Parser Performance by ~75%

Profiling the code showed that ~75% of the time was spent in the BufferedByteReader.pop() method. Almost all of that time was spent removing the read character from the buffer.

By storing an index, and removing elements from the buffer only when new elements are read, this bottleneck was removed.

Additionally, the peek(at:) method was rewritten to allow arbitraryly large lookaheads.

brutella commented on swift-csv/CSV.swift:
```diff
@@ -551,29 +551,26 @@ internal class BufferedByteReader {
-        
-        return nil
+    func peek(at peekIndex: Int) -> UInt8? {
+		while bufferIndex + peekIndex >= buffer.count {
```
Why use `while` here and not just `if`?

jakob replied:
`while` is necessary because read() might return fewer bytes than requested, even when we are not at the end of the file yet.

brutella commented on swift-csv/CSV.swift:
```diff
@@ -551,29 +551,26 @@ internal class BufferedByteReader {
-        return nil
+    func peek(at peekIndex: Int) -> UInt8? {
+		while bufferIndex + peekIndex >= buffer.count {
+			if read(bufferIndex + peekIndex + 1 - buffer.count) == 0 {
```
If we would change this to read a minimum number of bytes (e.g. 100),  we could call this from `pop()` as well.

```swift
// Read 100 bytes or more (if needed)
let readAmount = max(100, bufferIndex + peekIndex + 1 - buffer.count)
if read(readAmount) == 0 {
    return nil
}
```

And `pop()` would be like this.

```swift
func pop() -> UInt8? {
    guard let byte = self.peek() else {
        isAtEnd = true
        return nil
    }
    
    bufferIndex += 1
    return byte
}
```

jakob replied:
We could also just always read exactly 100 bytes, until we have enough. Then we can save a bunch of logic (and read() wouldn't need an argument anymore)

brutella replied:
I came up with the following.

```swift
func pop() -> UInt8? {
    guard let byte = self.peek() else {
        isAtEnd = true
        return nil
    }
    
    bufferIndex += 1
    return byte
}

func peek(at index: Int) -> UInt8? {
    let peekIndex = bufferIndex + index
    guard peekIndex < buffer.count else {
        if read(100) == 0 {
            return nil
        }
        
        return self.peek(at:index)
    }
    
    return buffer[peekIndex]
}
```

I find that easier to read.

jakob replied:
looks good to me!

brutella replied:
Could you please update your pull request accordingly?

brutella commented:
I like where you are going with these changes.

Thanks for contributing.

jakob commented:
The big missing part in that code is error handling. read() returns -1 on error.  I haven't tested it yet, but I'm almost sure that it will crash the parser (creating the range will fail)

brutella commented:
We could mark `peek()` and `pop()` to throw errors and then throw any stream errors like this

```swift
switch read(100) {
case 0:
    return nil
case -1:
    if let error = inputStream.streamError {
        throw error
    }
    return nil
default:
    return try self.peek(at:index)
}
```

jakob committed:
Refactored BufferedByteReader

jakob commented:
I added the changes as suggested. I also changed the `read()` method to not try to copy bytes in case of an error.

However, to be honest, I think that a while loop would be easier to understand than recursion:

```swift
	func peek(at index: Int = 0) -> UInt8? {
		while bufferIndex + index >= buffer.count {
			guard read(100) > 0 else {
				// end of file or error
				return nil
			}
		}
		return buffer[bufferIndex + index]
	}
```

instead of:

```swift
	func peek(at index: Int = 0) -> UInt8? {
		let peekIndex = bufferIndex + index
		guard peekIndex < buffer.count else {
			guard read(100) > 0 else {
				// end of file or error
				return nil
			}
			return self.peek(at:index)
		}
		return buffer[peekIndex]
	}
```

jakob commented:
I did not add real error handling yet, since that would affect other parts of the code.

brutella commented:
Looks good

brutella commented on swift-csv/CSV.swift:
```diff
@@ -532,34 +532,28 @@ internal class BufferedByteReader {
+	}
+	
+	/// - Returns: The character at `index`, or nil if the stream is at the end.
+	func peek(at index: Int = 0) -> UInt8? {
```
The default value is handy!
